### PR TITLE
Kubernetes and Vhive/firecracker-containerd

### PR DESCRIPTION
**Pull request by pogoba**
Try with 
```
sudo -E ./container-pid minio-deployment-877b8596f-kqdxr kubernetes
sudo -E ./container-pid default/helloworld-0-00001-deployment-56cf7467-rwpqn/user-container kubernetes
sudo -E ./container-pid knative-serving/autoscaler-589958b7b6-z5qfn kubernetes
sudo -E ./container-pid helloworld-0-00001-deployment-56cf7467-rwpqn vhive
```
For vhive you need this patch https://github.com/pogobanane/vhive/commit/ce450dce5490be4e41e5c77a9d18e8bd4875399d

Most dependencies are added though the url crate by dd27562 which could be removed again.

### CODE REVIEW

File: src/kubernetes.rs [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+        if path.is_dir() {
+            if let Ok(path) = visit_dirs(&path, containerdid) {
+                return Ok(path);
+            }
+        }
+    }
+    bail!("nothing found");
+}
+
+/// return any pid part of this cgroup
```
Comment (Mic92) [inline src/kubernetes.rs:157]: I suppose pid 1 will be the first pid in this cgroup, which is also what we usually want. So this is fine.

File: src/kubernetes.rs [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        .push("pods")
+        .push(pod_name);
+    let url = url.path();
+    let result = try_with!(
+        Command::new("kubectl")
+            .arg("get")
+            .arg("--raw")
+            .arg(url)
+            .output(),
+        "kubctl command cannot be spawned"
```
Comment (Mic92) [inline src/kubernetes.rs]:
```suggestion
        "kubectl command cannot be spawned"
```

File: src/kubernetes.rs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        bail!(
+            "kubectl get pod request failed (ret code {:?}): {}",
+            result.status.code(),
+            stderr
+        );
+    }
+
+    // parse response. Spec:
+    // https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#Pod
+    let json: json::Value = try_with!(
```
Comment (Mic92) [inline src/kubernetes.rs]: Serde takes awful long to compile. Do you mind using:

```
$ kubectl get pod -o jsonpath='{range .items[*].status.containerStatuses[*]}{.name}{"\t"}{.containerID}{"\n"}{end}'
runner  containerd://c6d16d9d996e23badc95bc374b86cd8284b9ef76f8650430c58bd28053896234
runner  containerd://eb3ff8ce29b2cca446db2f46b8fbee03dfe032087a8c4df993a39caf137f33ee
```

?

Should be easy to process using split?

File: src/vhive.rs [hunk 4]
```diff
@@ -0,0 +1,146 @@
+//! This module uses `kubectl` to get a containerd id, uses `journalctl` to match that id to a
+//! firecracker vm id and finally searches the open file-descriptors of relevant processes for one
+//! belonging to that vm id.
+//!
+//! The user may supply the same `container_id` as for the kubnetes module, the container always
+//! defaults to `user-container` because that is where the firecracker vm resides in vhive.
+//!
+//! Requires:
+//! - journald size longer than container lifetime
+//! - vhive with "debug log containerid->vmid" patch
```
Comment (Mic92) [inline src/vhive.rs:10]: I wonder if we could ask Dimitri to get an API endpoint for that... But works for now I guess.

File: src/kubernetes.rs (hunk 3 above)
Comment (pogoba) [reply src/kubernetes.rs]: yes 2c14a76

Comment (pogoba) [reply src/kubernetes.rs]: this PR does not add any deps anymore

File: src/kubernetes.rs (hunk 1 above)
Comment (pogoba) [reply src/kubernetes.rs:157]: > cgroup.procs: list of thread group IDs in the cgroup. This list is **not guaranteed to be sorted** or free of duplicate TGIDs, and userspace should sort/uniquify the list if this property is required. Writing a thread group ID into this file moves all threads in that group into this cgroup.

from https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v1/cgroups.html

Do we strictly require the cgroup's "root" pid?

File: src/vhive.rs (hunk 4 above)
Comment (pogoba) [reply src/vhive.rs:10]: the chances are good that i will add api endpoints to vhive. I could add this cid->vmid there.

File: src/kubernetes.rs (hunk 1 above)
Comment (Mic92) [reply src/kubernetes.rs:157]: The root pid has likely the most privileges. So eventually it would be better to have it.

**Pull request merged**